localstack/services/sns/publisher.py

```diff
@@ -0,0 +1,950 @@
+                "x-amz-sns-topic-arn": subscriber["TopicArn"],
+                "User-Agent": "Amazon Simple Notification Service Agent",
+            }
+            if message_context.type != "SubscriptionConfirmation":
+                # while testing, never had those from AWS but the docs above states it should be there
+                message_headers["x-amz-sns-subscription-arn"] = subscriber["SubscriptionArn"]
+
+            # When raw message delivery is enabled, x-amz-sns-rawdelivery needs to be set to 'true'
+            # indicating that the message has been published without JSON formatting.
+            # https://docs.aws.amazon.com/sns/latest/dg/sns-large-payload-raw-message-delivery.html
+            elif message_context.type == "Notification" and is_raw_message_delivery(subscriber):
+                message_headers["x-amz-sns-rawdelivery"] = "true"
+
+            response = requests.post(
+                subscriber["Endpoint"],
+                headers=message_headers,
+                data=message_body,
+                verify=False,
+            )
+
+            delivery = {
+                "statusCode": response.status_code,
+                "providerResponse": response.content.decode("utf-8"),
+            }
+            store_delivery_log(message_context, subscriber, success=True, delivery=delivery)
+
+            response.raise_for_status()
+        except Exception as exc:
+            LOG.info(
+                "Received error on sending SNS message, putting to DLQ (if configured): %s", exc
+            )
+            store_delivery_log(message_context, subscriber, success=False)
+            # AWS doesn't send to the DLQ if there's an error trying to deliver a UnsubscribeConfirmation msg
+            if message_context.type != "UnsubscribeConfirmation":
+                sns_error_to_dead_letter_queue(subscriber, message_body, str(exc))
+
+
+class EmailJsonTopicPublisher(BaseTopicPublisher):
+    def _publish(self, context: SnsPublishContext, subscriber: SnsSubscription):
+        ses_client = aws_stack.connect_to_service("ses")
+        if endpoint := subscriber.get("Endpoint"):
+            ses_client.verify_email_address(EmailAddress=endpoint)
+            ses_client.verify_email_address(EmailAddress="admin@localstack.com")
+            message_body = self.prepare_message(context.message, subscriber)
+            ses_client.send_email(
+                Source="admin@localstack.com",
+                Message={
+                    "Body": {"Text": {"Data": message_body}},
+                    "Subject": {"Data": "SNS-Subscriber-Endpoint"},
+                },
+                Destination={"ToAddresses": [endpoint]},
+            )
+            store_delivery_log(context.message, subscriber, success=True)
+
+
+class EmailTopicPublisher(EmailJsonTopicPublisher):
+    def prepare_message(self, message_context: SnsMessage, subscriber: SnsSubscription):
+        return message_context.message_content(subscriber["Protocol"])
```

Comment:
Just making sure here: Usually you would expect the hierarchy to go BaseClass -> Specific Class -> More Specific Class. So BaseTopicPublisher -> EmailTopicPublisher -> EmailJsonTopicPublisher. It seems a bit strange that the more generically named class gets all the special functionality of the more specific class, and then overwrites something else.

Reply:
Yes, this is a weird because of the naming.
The thing is there is some behaviour implemented in the base class because it's used by most of the publishers.
EmailJson is closer in behaviour to the other publishers because it formats the payload (done by default in `prepare_message`).
I think the naming is the issue, but I'm not sure how I should rename the `EmailTopicPublisher`?
I guess the "cleanest" would be to have a `BaseEmailTopicPublisher` implementing the `_publish` method, and have `EmailJsonTopicPublisher` inherits from `TopicPublisher` and `BaseEmailTopicPublisher` with no overriding of methods while `EmailTopicPublisher` would also inherits from both but override `prepare_message`.

Honestly, not sure what is the best way 😞

Reply:
Hm, tricky.
Another way could be to write `EmailTopicPublisher` so that it holds all the code, and then let `EmailTopicPublisher` call the original method.
```
class EmailTopicPublisher(TopicPublisher):
    def _publish(self, context: SnsPublishContext, subscriber: SnsSubscription):
          ...
    def prepare_message(self, message_context: SnsMessage, subscriber: SnsSubscription):
        ...

class EmailJsonTopicPublisher(EmailTopicPublisher):
    def prepare_message(self, message_context: SnsMessage, subscriber: SnsSubscription):
        # call the "super super" aka original method
        super(EmailTopicPublisher, self).prepare_message(message_context, subscriber)
```
DISCLAIMER: I have no idea if this is pretty, or good practice, or very pythonic. I literally just looked it up, and I don' think I ever called a grandparents method before. In ANY language. The main purpose would be to avoid an overly complicated class hierarchy for the sake of the hierarchy (all that effort for 2 methods total seems overkill).

It probably comes down to whether
1) `EmailJsonTopicPublisher` is actually a specialized instance of `EmailTopicPublisher`, as suggested by the name
2) it is like you coded it, the other way around (then we should change the name)
3) They are not "parent/child" but siblings, which would be your new suggested code
from a semantic point of view.
I would ping @thrau for input, but then again we might be blowing this out of proportion. The classes a very small, so a simple comment might solve this too.

Reply:
I think this is now sufficiently addressed by the added documentation